### PR TITLE
#165413097 Setup Api Documentation

### PR DESCRIPTION
**Pull request by Niyitangasam**
Description
=======
The ability of API is to describe its usage to the consumers. This PR set up API documentation using swagger and we will continue to update it throughout the development process.

Type of change
=======
N/A

How has it been tested
=======
It can be tested through  `/docs` appended on the base URL of the hosting server

Checklist:
=======
N/A

Pivotal tracker story ID
=======
[165413097](https://www.pivotaltracker.com/n/projects/2326984/stories/165413097)

### CODE REVIEW

**Review by AJAkimana — APPROVED** [review]
Great work @Niyitangasam.

**Comment by Kabalisa** [comment]
Good Job @Niyitangasam. swagger is set up well all you can do is add endpoint documentations as we do them.

File: index.js [hunk 1]
```diff
@@ -1,93 +1,53 @@
...
-} else {
-    mongoose.connect("mongodb://localhost/conduit");
-    mongoose.set("debug", true);
-}
+// initialize swaggerJSDoc
+const swaggerSpec = swaggerJSDoc(options);
 
-require("./models/User");
+// route for swagger.json
+app.get('/swagger.json', function(req, res) {
```
Comment (macphilips) [inline index.js]: what is the purpose of this route?

File: package.json [hunk 2]
```diff
@@ -4,29 +4,34 @@
...
-    "method-override": "^2.3.10",
-    "methods": "^1.1.2",
     "mongoose": "^5.2.2",
     "mongoose-unique-validator": "^2.0.1",
-    "morgan": "^1.9.0",
     "passport": "^0.4.0",
     "passport-local": "^1.0.0",
     "request": "^2.87.0",
-    "slug": "^0.9.1",
+    "save": "^2.3.3",
```
Comment (macphilips) [inline package.json]: What are you using this package for?

File: index.js [hunk 3]
```diff
@@ -1,93 +1,53 @@
...
+import swaggerui from 'swagger-ui-express';
+//import swaggerDoc from './docs/swagger';
 
-const isProduction = process.env.NODE_ENV === "production";
 
 // Create global app object
 const app = express();
 
-app.use(cors());
+// swagger definition
```
Comment (macphilips) [inline index.js]: Can you define the swagger configuration in a separate file and add it to a config folder

File: index.js [hunk 4]
```diff
@@ -1,93 +1,53 @@
...
-    })
-);
-
-if (!isProduction) {
-    app.use(errorhandler());
-}
+// options for swagger jsdoc 
+const options = {
+    swaggerDefinition: swaggerDefinition,
+    apis: ['./routes/*.js']
```
Comment (macphilips) [inline index.js]: update the pattern to recursive search any js file within the directory:
```suggestion
    apis: ['./routes/**/*.js']
```

File: index.js (hunk 3 above)
Comment (macphilips) [reply index.js]: Also, add sample documentation, so your peers know how to you it

File: index.js [hunk 5]
```diff
@@ -1,93 +1,53 @@
...
 
-app.use(cors());
+// swagger definition
+const swaggerDefinition = {
+    info: {
+      title: 'Authors Haven',
+      version: '1.0.0',
+      description: 'A Social platform for the creative at heart',
+    },
+    host: '',
```
Comment (macphilips) [inline index.js]: You need to make the host field configurable by using environment variables. You can do this by using [dotenv](https://www.npmjs.com/package/dotenv) package.
```suggestion
    host: `${process.env.BASE_URL}`,
```

File: index.js [hunk 6]
```diff
@@ -1,93 +1,53 @@
...
+    basePath: '/',
```
Comment (macphilips) [inline index.js]:
```suggestion
    basePath: '/api/v1',
```

**Review by macphilips — CHANGES_REQUESTED** [review]
You've done a nice job so far, but this task is not complete. I have made some comments kindly attend to it.

**Pull request merged**